modules/core/src/main/scala/fs2/kafka/internal/KafkaConsumerActor.scala

```diff
@@ -135,13 +135,36 @@ private[kafka] final class KafkaConsumerActor[F[_], K, V](
     res.start.void
   }
 
-  private[this] def assigned(assigned: SortedSet[TopicPartition]): F[Unit] =
-    ref
-      .updateAndGet(_.withRebalancing(false))
-      .flatMap { state =>
-        log(AssignedPartitions(assigned, state)) >>
-          state.onRebalances.foldLeft(F.unit)(_ >> _.onAssigned(assigned))
+  private[this] def assigned(assigned: SortedSet[TopicPartition]): F[Unit] = {
+    def withState[A] = StateT.apply[Id, State[F, K, V], A](_)
+
+    val removeRevokedRecords = withState { st =>
+      val assignments = st.assignments
+      val revokedRecords = st.records.filterKeysStrict(!assignments(_))
+
+      if (revokedRecords.nonEmpty) {
+        val newState = st.withoutRecords(revokedRecords.keySet)
+
+        val action = logging.log(RemovedRevokedRecords(revokedRecords, newState))
+
+        (newState, action)
+      } else (st, F.unit)
+    }
+
+    for {
+      action <- ref.modify { state =>
+        removeRevokedRecords.run(
+          state
+            .withRebalancing(false)
+            .withAssignments(assigned)
+        )
       }
+      updatedState <- ref.get
+      _ <- action >>
+        log(AssignedPartitions(assigned, updatedState)) >>
+        updatedState.onRebalances.foldLeft(F.unit)(_ >> _.onAssigned(assigned))
+    } yield ()
```

Comment:
Instead of a separate `ref.get` can we return the new state from `ref.modify`:
```suggestion
    ref
      .modify { state =>
        removeRevokedRecords
          .run(
            state
              .withRebalancing(false)
              .withAssignments(assigned)
          )
          .map(stateAndAction => stateAndAction._1 -> stateAndAction)
      }
      .flatMap {
        case (updatedState, action) =>
          action >> log(AssignedPartitions(assigned, updatedState)) >>
            updatedState.onRebalances.foldLeft(F.unit)(_ >> _.onAssigned(assigned))
      }
```

Reply:
Yeah, I didn't notice it could be done this way.
Great thanks! 👍